content/posts/GSoC_2021_Quarter/index.md

```diff
@@ -0,0 +1,144 @@
+
+Anyway, if you already know what Faces are, it might strike you:
+
+If we already have all the faces we need from multiple fonts (let's say we created a child of FT2Font.. which only <ins>tracks the faces</ins> for its families), we should be able to render everything from that parent FT2Font right?
+
+As I later figured out while finding segfaults in implementing this design:
+```
+Each FT2Font is linked to a single FT_Library object!
+```
+
+If you tried to load the face/glyph/character (basically anything) from a different FT2Font object.. you'll run into serious segfaults. (because one object linked to an `FT_Library` can't really access another object which has it's own `FT_Library`)
+```cpp
+// face is linked to FT2Font; which is
+// linked to a single FT_Library object
+FT_Face face = this->get_face();
+FT_Get_Glyph(face->glyph, &placeholder); // works like a charm
+
+// somehow get another FT2Font's face
+FT_Face family_face = this->get_family_member()->get_face();
+FT_Get_Glyph(family_face->glyph, &placeholder); // segfaults!
+```
+
+Realizing this took a good amount of time! After this I quickly came up with a recursive approach, wherein we:
+1. Create a list of FT2Font objects within Python, and pass it down to FT2Font
+2. FT2Font will hold pointers to its families via a \
+		`std::vector<FT2Font *> fallback_list`
+3. Find if the character we want is available in the current font
+    1. If the character is available, use that FT2Font to render that character
+    2. If the character isn't found, go to step 3 again, but now iterate through the `fallback_list`
+4. That's it!
+
+A quick overhaul of the above piece of code^
+```cpp
+bool ft_get_glyph(FT_Glyph &placeholder) {
+	FT_Error not_found = FT_Get_Glyph(this->get_face(), &placeholder);
+	if (not_found) return False;
+	else return True;
+}
+
+// within driver code
+for (uint i=0; i<fallback_list.size(); i++) {
+	// iterate through all FT2Font objects
+	bool was_found = fallback_list[i]->ft_get_glyph(placeholder);
+	if (was_found) break;
+}
+```
+
+With the idea surrounding this implementation, the [Agg backend](https://matplotlib.org/stable/api/backend_agg_api.html) is able to render a document (either through GUI, or a PNG) with multiple fonts!
+
+<p align="center">
+    <figure>
+        <img src="https://user-images.githubusercontent.com/43996118/128347495-1f4f858d-33d3-4119-8732-5b26c4e9ca2a.png" alt="ChineseInBetween" />
+        <figcaption style="text-align: center; font-style: italic;">PNG straight outta Matplotlib!</figcaption>
+    </figure>
+</p>
+
+## Python C-API is hard, at first!
+I've spent days at Python C-API's [argument doc](https://docs.python.org/3/c-api/arg.html), and it's hard to get what you need at first, ngl.
+
+But, with the help of some amazing people in GSoC community ([@srijan-paul](https://srijan-paul.github.io/), [@atharvaraykar](https://atharvaraykar.me/)) and amazing mentors, blockers begone!

```

Comment:
```suggestion
But, with the help of some amazing people in the GSoC community ([@srijan-paul](https://srijan-paul.github.io/), [@atharvaraykar](https://atharvaraykar.me/)) and amazing mentors, blockers begone!
```

also 😁 that's so awesome!

Reply:
also to remind myself for when this gets tweeted out, @_injuly & @AtharvaRaykar

Reply:
😄 Cool